### DISCUSSION TITLE
xdg-terminal-exec support

### DISCUSSION THREAD

**Discussion by ManuelSchneid3r** [linuxdeepin/developer-center#9817] [Features Request & Ideas | 特性请求 & 头脑风暴]
Running a command in a terminal from another app is a common usecase. Almost all terminals provide such a command execution interface (`-e`, `-x`, `--` or such). Clients can [guess](https://github.com/albertlauncher/albert/blob/6783257df7e15dcfbae49d03023414b7c3692dbd/src/app/terminalprovider.cpp#L120-L150), but they cant know it for sure. If it *can* break, it *will* break.

There is a [freedesktop spec proposal](https://gitlab.freedesktop.org/terminal-wg/specifications/-/merge_requests/3) that tries to standardize the way we configure a default terminal for command execution. It implicitly solves the above problem. Its as simple as that: the app/package provides this info.

To support it you simply need to add `X-ExecArgs` or `ExecArgs` ([the reference implementation respects both](https://github.com/Vladimir-csp/xdg-terminal-exec/blob/af3570e1af8691416e77171e7e29a03572d541c4/xdg-terminal-exec#L387)) to the desktop file ([like this](https://github.com/GNOME/gnome-terminal/blob/f1814cb7ecfe6a5cb4f629dcea56db0dec10a242/data/org.gnome.Terminal.desktop.in#L14)) and make sure there is `TerminalEmulator` in `Categories`. It has not yet been approved (therefore the `X-`) but hey `ExecArgs` does not conflict any standard key and if it will be approved you dont have to change it again.

As an elaborative example see how GNOME terminal does it: https://github.com/GNOME/gnome-terminal/commit/f1814cb7ecfe6a5cb4f629dcea56db0dec10a242

This will help making Linux a better user experience and a less annoying developer experience.

Thank you

PS:
The draft states that the default will be `-e` and you might think "okay, done" if your ExecArg *is* `-e`. I recommend to add it anyway because then you are on the safe side if the final spec does not fall back to `-e`. I guess (hope) there will be no default. 
You might also think "Why that early if it's a draft?". Correct, but if you do so, I will give you a grateful hug should we ever meet in real life because I am the maintainer of that frustrating list linked in the beginning. If you do so I can wipe your entry from the list :heart:. Also this proposal is a good thing and we should all push it.

**Comment by ManuelSchneid3r** [comment]
You can see what others are doing [here](https://github.com/albertlauncher/albert-plugin-applications/issues/1)

**Comment by BLumia** [comment]
I am a little bit confused about the existing implementations. So if I understand it correctly, adding `X-TerminalArgExec=-e` to deepin-terminal could be enough?

For example, in [your list](https://github.com/albertlauncher/albert-plugin-applications/issues/1), [GNOME Terminal uses `X-ExecArgs=--`](https://github.com/GNOME/gnome-terminal/blob/8cb21c53bea7ebd5e4e6cad34e6f072fadd61c66/data/org.gnome.Terminal.desktop.in#L14), but the actual proposal suggest to use `X-TerminalArgExec=--`. It seems [GNOME Terminal also support the later one](https://github.com/GNOME/gnome-terminal/blob/5ba7dfb204848043cbdc5a8a489ee025dc55477f/data/org.gnome.Terminal.desktop.in#L15) in most recent commit tho, introduced in https://github.com/GNOME/gnome-terminal/commit/d4285151c618e3f75e0041de9641af4dc1bb4c4b.

**Reply by ManuelSchneid3r** [reply]
Well it is an evolving [spec](https://github.com/Vladimir-csp/xdg-terminal-exec#desktop-entry-for-a-terminal). I'd personally prefer if they settle to a single `TerminalExec` line. Things could still change. But we need the feedback from the field.

**Reply by BLumia** [reply]
See:

- https://github.com/linuxdeepin/deepin-terminal/pull/555
